File: sys/src/linear/noise_model.cc
#include "noise_model.h"
#include "../base/rust.hpp"

namespace gtsam {

std::shared_ptr<DiagonalNoiseModel> from_diagonal_noise_model_sigmas(const rust::Slice<double> sigmas) {
  Vector eigenSigmas = Eigen::Map<Vector>(sigmas.data(), sigmas.size(), 1);
  constexpr bool smart = true;

  return to_std_ptr(DiagonalNoiseModel::Sigmas(eigenSigmas, smart));
}

std::shared_ptr<BaseNoiseModel> cast_diagonal_noise_model_to_base_noise_model(
    const std::shared_ptr<DiagonalNoiseModel> &a) {
  return downcast<DiagonalNoiseModel, BaseNoiseModel>(a);
}

std::shared_ptr<IsotropicNoiseModel> from_isotropic_noise_model_dim_and_sigma(const size_t dim, const double sigma) {
    constexpr bool smart = true;

    return to_std_ptr(IsotropicNoiseModel::Sigma(dim, sigma, smart));
}

std::shared_ptr<BaseNoiseModel> cast_isotropic_noise_model_to_base_noise_model(
    const std::shared_ptr<IsotropicNoiseModel> &a) {
    return downcast<IsotropicNoiseModel, BaseNoiseModel>(a);
}

} // namespace gtsam

File: src/linear/noise_model.rs
use cxx::SharedPtr;
use nalgebra::SVector;

pub struct BaseNoiseModel {
    pub(crate) inner: SharedPtr<::sys::BaseNoiseModel>,
}

pub struct DiagonalNoiseModel {
    pub(crate) inner: SharedPtr<::sys::DiagonalNoiseModel>,
}

impl DiagonalNoiseModel {
    pub fn from_sigmas<const D: usize>(mut sigmas: SVector<f64, D>) -> Self {
        Self {
            inner: ::sys::from_diagonal_noise_model_sigmas(sigmas.data.as_mut_slice()),
        }
    }

    pub(crate) fn to_base_model(&self) -> BaseNoiseModel {
        BaseNoiseModel {
            inner: ::sys::cast_diagonal_noise_model_to_base_noise_model(&self.inner),
        }
    }
}

pub struct IsotropicNoiseModel {
    pub(crate) inner: SharedPtr<::sys::IsotropicNoiseModel>,
}

impl IsotropicNoiseModel {
    pub fn from_dim_and_sigma(dim: usize, sigma: f64) -> Self {
        Self {
            inner: ::sys::from_isotropic_noise_model_dim_and_sigma(dim, sigma),
        }
    }

    pub(crate) fn to_base_model(&self) -> BaseNoiseModel {
        BaseNoiseModel {
            inner: ::sys::cast_isotropic_noise_model_to_base_noise_model(&self.inner),
        }
    }
}
